File: backend/app/core/simulator.py
from __future__ import annotations

from typing import Any
from app.core.geometry import snapshot
from app.core.routing import (
    RoutingMetric,
    build_adjacency,
    classify_failure,
    find_route,
)


def run_simulation(
    scenario: dict[str, Any],
    metric: RoutingMetric = "hops",
    include_timeline: bool = True,
) -> dict[str, Any]:
    """
    Run full time-horizon simulation on scenario.
    Returns complete metrics, outage timelines, and failure breakdown.
    """
    env = scenario["environment"]
    step_s = int(env["step_s"])
    horizon_s = int(env["horizon_s"])
    target_avail = float(env.get("target_availability", 0.9))

    clients = [g["id"] for g in scenario.get("ground_sites", []) if g.get("role") == "client"]
    all_clients_set = set(clients)
    gateways = [g["id"] for g in scenario.get("ground_sites", []) if g.get("role") == "gateway"]
    all_gw_set = set(gateways)
    gw_outages = scenario.get("gateway_outages", [])

    time_steps = list(range(0, horizon_s, step_s))
    total_steps = len(time_steps)

    # Per-client accumulators
    client_data: dict[str, dict[str, Any]] = {
        cid: {
            "steps_with_path": 0,
            "steps_with_vis": 0,
            "current_outage_steps": 0,
            "max_outage_steps": 0,
            "hops_list": [],
            "distance_list": [],
            "failure_counts": {
                "no_client_satellite": 0,
                "gateway_offline": 0,
                "no_gateway_satellite": 0,
                "isl_disconnected": 0,
            },
            "timeline": [],
            "current_outage_start": None,
            "current_outage_reason": None,
            "outage_intervals": [],
        }
        for cid in clients
    }

    all_routes: list[dict[str, Any]] = []

    for t_s in time_steps:
        snap = snapshot(scenario, t_s, fast_edges_only=True)
        adj = build_adjacency(snap["edges"])

        # Online gateways at this time step
        cur_gw_outages = {f["gateway_id"] for f in gw_outages if f["start_s"] <= t_s < f["end_s"]}
        online_gateways = {gid for gid in all_gw_set if gid not in cur_gw_outages}

        # Any active satellite visible to any online gateway?
        gw_has_sat = False
        for gid in online_gateways:
            if any(nxt not in all_gw_set for nxt, _ in adj.get(gid, [])):
                gw_has_sat = True
                break

        for cid in clients:
            cdata = client_data[cid]
            client_visible_sats = [
                nxt
                for nxt, _ in adj.get(cid, [])
                if nxt not in all_clients_set and nxt not in all_gw_set
            ]
            has_client_sat = len(client_visible_sats) > 0

            if has_client_sat:
                cdata["steps_with_vis"] += 1

            path, total_dist = find_route(
                adj=adj,
                client_id=cid,
                online_gateways=online_gateways,
                all_clients=all_clients_set,
                metric=metric,
            )

            all_routes.append(
                {
                    "t_s": t_s,
                    "client_id": cid,
                    "path": path,
                }
            )

            has_path = len(path) > 0

            if has_path:
                cdata["steps_with_path"] += 1
                hops = len(path) - 1
                cdata["hops_list"].append(hops)
                cdata["distance_list"].append(total_dist)

                # Close ongoing outage interval if one was active
                if cdata["current_outage_steps"] > 0:
                    outage_dur = cdata["current_outage_steps"] * step_s
                    cdata["outage_intervals"].append(
                        {
                            "start_s": cdata["current_outage_start"],
                            "end_s": t_s,
                            "duration_s": outage_dur,
                            "reason": cdata["current_outage_reason"],
                        }
                    )
                    cdata["current_outage_steps"] = 0
                    cdata["current_outage_start"] = None
                    cdata["current_outage_reason"] = None

                if include_timeline:
                    cdata["timeline"].append(
                        {
                            "t_s": t_s,
                            "status": "ok",
                            "has_route": True,
                            "has_visibility": has_client_sat,
                            "hops": hops,
                            "path": path,
                            "distance_km": round(total_dist, 2),
                            "failure_code": None,
                            "failure_reason": None,
                        }
                    )
            else:
                fail_code, fail_desc = classify_failure(
                    has_client_satellite=has_client_sat,
                    online_gateways_count=len(online_gateways),
                    has_gateway_satellite=gw_has_sat,
                )
                cdata["failure_counts"][fail_code] += 1

                if cdata["current_outage_steps"] == 0:
                    cdata["current_outage_start"] = t_s
                    cdata["current_outage_reason"] = fail_code

                cdata["current_outage_steps"] += 1
                if cdata["current_outage_steps"] > cdata["max_outage_steps"]:
                    cdata["max_outage_steps"] = cdata["current_outage_steps"]

                if include_timeline:
                    cdata["timeline"].append(
                        {
                            "t_s": t_s,
                            "status": "outage",
                            "has_route": False,
                            "has_visibility": has_client_sat,
                            "hops": 0,
                            "path": [],
                            "distance_km": 0.0,
                            "failure_code": fail_code,
                            "failure_reason": fail_desc,
                        }
                    )

    # Close any trailing outage interval at horizon_s
    for cid in clients:
        cdata = client_data[cid]
        if cdata["current_outage_steps"] > 0:
            outage_dur = cdata["current_outage_steps"] * step_s
            cdata["outage_intervals"].append(
                {
                    "start_s": cdata["current_outage_start"],
                    "end_s": horizon_s,
                    "duration_s": outage_dur,
                    "reason": cdata["current_outage_reason"],
                }
            )

    # Format final metrics
    client_metrics: dict[str, Any] = {}
    total_avail_pct = 0.0
    min_avail_pct = 100.0

    for cid in clients:
        cdata = client_data[cid]
        avail_ratio = cdata["steps_with_path"] / total_steps if total_steps > 0 else 0.0
        vis_ratio = cdata["steps_with_vis"] / total_steps if total_steps > 0 else 0.0
        avail_pct = avail_ratio * 100.0
        vis_pct = vis_ratio * 100.0

        total_avail_pct += avail_pct
        min_avail_pct = min(min_avail_pct, avail_pct)

        max_outage_s = cdata["max_outage_steps"] * step_s
        total_outage_s = (total_steps - cdata["steps_with_path"]) * step_s

        hops = cdata["hops_list"]
        mean_hops = float(sum(hops) / len(hops)) if hops else None
        max_hops = max(hops) if hops else None
        min_hops = min(hops) if hops else None

        dists = cdata["distance_list"]
        mean_dist = float(sum(dists) / len(dists)) if dists else None

        client_metrics[cid] = {
            "client_id": cid,
            "availability_pct": round(avail_pct, 2),
            "availability_ratio": round(avail_ratio, 4),
            "visibility_pct": round(vis_pct, 2),
            "visibility_ratio": round(vis_ratio, 4),
            "max_outage_s": max_outage_s,
            "total_outage_s": total_outage_s,
            "mean_hops": round(mean_hops, 2) if mean_hops is not None else None,
            "max_hops": max_hops,
            "min_hops": min_hops,
            "mean_distance_km": round(mean_dist, 2) if mean_dist is not None else None,
            "meets_target": avail_ratio >= target_avail,
            "target_availability": target_avail,
            "failure_breakdown": cdata["failure_counts"],
            "outage_intervals": cdata["outage_intervals"],
            "timeline": cdata["timeline"] if include_timeline else [],
        }

    avg_avail = round(total_avail_pct / len(clients), 2) if clients else 0.0

    return {
        "scenario_meta": scenario.get("meta", {}),
        "total_steps": total_steps,
        "step_s": step_s,
        "horizon_s": horizon_s,
        "target_availability": target_avail,
        "summary": {
            "average_availability_pct": avg_avail,
            "min_availability_pct": round(min_avail_pct, 2) if clients else 0.0,
            "all_meet_target": all(m["meets_target"] for m in client_metrics.values()),
        },
        "client_metrics": client_metrics,
        "routes": all_routes,
    }


simulate_scenario = run_simulation

__all__ = ["run_simulation", "simulate_scenario"]

File: backend/crates/cosmo-core/src/simulator.rs
use std::collections::{HashMap, HashSet};

use crate::geometry::snapshot;
use crate::models::{
    ClientMetrics, OutageInterval, RouteRecord, Scenario, SimulationResult, SimulationSummary,
    TimelineEntry,
};
use crate::routing::{build_adjacency, classify_failure, find_route, RoutingMetric};

fn round2(val: f64) -> f64 {
    (val * 100.0).round() / 100.0
}

fn round4(val: f64) -> f64 {
    (val * 10000.0).round() / 10000.0
}

struct InternalClientData {
    steps_with_path: usize,
    steps_with_vis: usize,
    current_outage_steps: usize,
    max_outage_steps: usize,
    hops_list: Vec<usize>,
    distance_list: Vec<f64>,
    failure_counts: HashMap<String, usize>,
    timeline: Vec<TimelineEntry>,
    current_outage_start: Option<f64>,
    current_outage_reason: Option<String>,
    outage_intervals: Vec<OutageInterval>,
}

pub fn run_simulation(
    scenario: &Scenario,
    metric: RoutingMetric,
    include_timeline: bool,
) -> SimulationResult {
    let env = &scenario.environment;
    let step_s = env.step_s;
    let horizon_s = env.horizon_s;
    let target_avail = env.target_availability;

    let clients: Vec<String> = scenario
        .ground_sites
        .iter()
        .filter(|g| g.role == "client")
        .map(|g| g.id.clone())
        .collect();
    let all_clients_set: HashSet<String> = clients.iter().cloned().collect();

    let gateways: Vec<String> = scenario
        .ground_sites
        .iter()
        .filter(|g| g.role == "gateway")
        .map(|g| g.id.clone())
        .collect();
    let all_gw_set: HashSet<String> = gateways.iter().cloned().collect();

    let time_steps: Vec<i64> = (0..horizon_s).step_by(step_s as usize).collect();
    let total_steps = time_steps.len();

    let mut client_data: HashMap<String, InternalClientData> = clients
        .iter()
        .map(|cid| {
            let mut fc = HashMap::new();
            fc.insert("no_client_satellite".to_string(), 0);
            fc.insert("gateway_offline".to_string(), 0);
            fc.insert("no_gateway_satellite".to_string(), 0);
            fc.insert("isl_disconnected".to_string(), 0);

            (
                cid.clone(),
                InternalClientData {
                    steps_with_path: 0,
                    steps_with_vis: 0,
                    current_outage_steps: 0,
                    max_outage_steps: 0,
                    hops_list: vec![],
                    distance_list: vec![],
                    failure_counts: fc,
                    timeline: vec![],
                    current_outage_start: None,
                    current_outage_reason: None,
                    outage_intervals: vec![],
                },
            )
        })
        .collect();

    let mut all_routes: Vec<RouteRecord> = Vec::new();

    for &t_step in &time_steps {
        let t_s = t_step as f64;
        let snap = snapshot(scenario, t_s, true);
        let adj = build_adjacency(&snap.edges);

        let cur_gw_outages: HashSet<&str> = scenario
            .gateway_outages
            .iter()
            .filter(|f| f.start_s <= t_s && t_s < f.end_s)
            .map(|f| f.gateway_id.as_str())
            .collect();

        let online_gateways: HashSet<String> = all_gw_set
            .iter()
            .filter(|gid| !cur_gw_outages.contains(gid.as_str()))
            .cloned()
            .collect();

        let mut gw_has_sat = false;
        for gid in &online_gateways {
            if let Some(neighbors) = adj.get(gid) {
                if neighbors.iter().any(|(nxt, _)| !all_gw_set.contains(nxt)) {
                    gw_has_sat = true;
                    break;
                }
            }
        }

        for cid in &clients {
            let cdata = client_data.get_mut(cid).unwrap();

            let has_client_sat = adj.get(cid).map_or(false, |neighbors| {
                neighbors
                    .iter()
                    .any(|(nxt, _)| !all_clients_set.contains(nxt) && !all_gw_set.contains(nxt))
            });

            if has_client_sat {
                cdata.steps_with_vis += 1;
            }

            let (path, total_dist) = find_route(&adj, cid, &online_gateways, &all_clients_set, metric);

            all_routes.push(RouteRecord {
                t_s,
                client_id: cid.clone(),
                path: path.clone(),
            });

            let has_path = !path.is_empty();

            if has_path {
                cdata.steps_with_path += 1;
                let hops = path.len() - 1;
                cdata.hops_list.push(hops);
                cdata.distance_list.push(total_dist);

                if cdata.current_outage_steps > 0 {
                    let outage_dur = (cdata.current_outage_steps * step_s as usize) as f64;
                    cdata.outage_intervals.push(OutageInterval {
                        start_s: cdata.current_outage_start.unwrap(),
                        end_s: t_s,
                        duration_s: outage_dur,
                        reason: cdata.current_outage_reason.take().unwrap(),
                    });
                    cdata.current_outage_steps = 0;
                    cdata.current_outage_start = None;
                }

                if include_timeline {
                    cdata.timeline.push(TimelineEntry {
                        t_s,
                        status: "ok".to_string(),
                        has_route: true,
                        has_visibility: has_client_sat,
                        hops,
                        path,
                        distance_km: round2(total_dist),
                        failure_code: None,
                        failure_reason: None,
                    });
                }
            } else {
                let (fail_code, fail_desc) =
                    classify_failure(has_client_sat, online_gateways.len(), gw_has_sat);
                *cdata.failure_counts.get_mut(fail_code).unwrap() += 1;

                if cdata.current_outage_steps == 0 {
                    cdata.current_outage_start = Some(t_s);
                    cdata.current_outage_reason = Some(fail_code.to_string());
                }

                cdata.current_outage_steps += 1;
                if cdata.current_outage_steps > cdata.max_outage_steps {
                    cdata.max_outage_steps = cdata.current_outage_steps;
                }

                if include_timeline {
                    cdata.timeline.push(TimelineEntry {
                        t_s,
                        status: "outage".to_string(),
                        has_route: false,
                        has_visibility: has_client_sat,
                        hops: 0,
                        path: vec![],
                        distance_km: 0.0,
                        failure_code: Some(fail_code.to_string()),
                        failure_reason: Some(fail_desc.to_string()),
                    });
                }
            }
        }
    }

    for cid in &clients {
        let cdata = client_data.get_mut(cid).unwrap();
        if cdata.current_outage_steps > 0 {
            let outage_dur = (cdata.current_outage_steps * step_s as usize) as f64;
            cdata.outage_intervals.push(OutageInterval {
                start_s: cdata.current_outage_start.unwrap(),
                end_s: horizon_s as f64,
                duration_s: outage_dur,
                reason: cdata.current_outage_reason.take().unwrap(),
            });
        }
    }

    let mut client_metrics = HashMap::new();
    let mut total_avail_pct = 0.0;
    let mut min_avail_pct = 100.0;

    for cid in &clients {
        let cdata = client_data.get_mut(cid).unwrap();
        let avail_ratio = if total_steps > 0 {
            cdata.steps_with_path as f64 / total_steps as f64
        } else {
            0.0
        };
        let vis_ratio = if total_steps > 0 {
            cdata.steps_with_vis as f64 / total_steps as f64
        } else {
            0.0
        };
        let avail_pct = avail_ratio * 100.0;
        let vis_pct = vis_ratio * 100.0;

        total_avail_pct += avail_pct;
        if avail_pct < min_avail_pct {
            min_avail_pct = avail_pct;
        }

        let max_outage_s = (cdata.max_outage_steps * step_s as usize) as f64;
        let total_outage_s = ((total_steps - cdata.steps_with_path) * step_s as usize) as f64;

        let mean_hops = if !cdata.hops_list.is_empty() {
            let sum: usize = cdata.hops_list.iter().sum();
            Some(round2(sum as f64 / cdata.hops_list.len() as f64))
        } else {
            None
        };
        let max_hops = cdata.hops_list.iter().copied().max();
        let min_hops = cdata.hops_list.iter().copied().min();

        let mean_dist = if !cdata.distance_list.is_empty() {
            let sum: f64 = cdata.distance_list.iter().sum();
            Some(round2(sum / cdata.distance_list.len() as f64))
        } else {
            None
        };

        client_metrics.insert(
            cid.clone(),
            ClientMetrics {
                client_id: cid.clone(),
                availability_pct: round2(avail_pct),
                availability_ratio: round4(avail_ratio),
                visibility_pct: round2(vis_pct),
                visibility_ratio: round4(vis_ratio),
                max_outage_s,
                total_outage_s,
                mean_hops,
                max_hops,
                min_hops,
                mean_distance_km: mean_dist,
                meets_target: avail_ratio >= target_avail,
                target_availability: target_avail,
                failure_breakdown: cdata.failure_counts.clone(),
                outage_intervals: cdata.outage_intervals.clone(),
                timeline: if include_timeline {
                    cdata.timeline.clone()
                } else {
                    vec![]
                },
            },
        );
    }

    let avg_avail = if !clients.is_empty() {
        round2(total_avail_pct / clients.len() as f64)
    } else {
        0.0
    };

    let summary = SimulationSummary {
        average_availability_pct: avg_avail,
        min_availability_pct: if !clients.is_empty() {
            round2(min_avail_pct)
        } else {
            0.0
        },
        all_meet_target: client_metrics.values().all(|m| m.meets_target),
    };

    SimulationResult {
        scenario_meta: scenario.meta.clone().unwrap_or(serde_json::Value::Null),
        total_steps,
        step_s,
        horizon_s,
        target_availability: target_avail,
        summary,
        client_metrics,
        routes: all_routes,
    }
}

pub fn simulate_scenario(
    scenario: &Scenario,
    metric: RoutingMetric,
    include_timeline: bool,
) -> SimulationResult {
    run_simulation(scenario, metric, include_timeline)
}
